packages/actor-rdf-resolve-hypermedia-links-queue-priority/lib/LinkQueuePriority.ts

```diff
@@ -0,0 +1,229 @@
+export class LinkQueuePriority implements ILinkQueue {
+  /**
+   * Max heap with links and their priorities
+   */
+  public readonly links: ILink[] = [];
+  /**
+   * Data structure to track URLs in the queue and allow for constant time lookup of links in queue based
+   * on link URL
+   */
+  public readonly urlToLink: Record<string, ILink> = {};
+
+  /**
+   * Pushes element to heap by appending it to array and up-heaping the new element
+   */
+  public push(link: ILink): boolean {
+    const idx: number = this.links.length;
+
+    // If we push a link that has no metadata or has metadata but no priority we set priority to 0
+    // and always set index to end of array (and upheap from there)
+    if (!link.metadata || !link.metadata.priority) {
+      link.metadata = { priority: 0, index: idx };
+    } else {
+      link.metadata.index = idx;
+    }
+    this.links.push(link);
+
+    // Add to Records to allow fast updates in priority
+    this.urlToLink[link.url] = link;
+    this.upHeap(idx);
+    return true;
+  }
+
+  /**
+   * Pops the highest priority link and returns it. Then it sets the last element as root and
+   * down-heaps it.
+   * @returns popped element
+   */
+  public pop(): ILink {
+    const max = this.links[0];
+    const endArray = this.links.pop();
+
+    // If we remove link, remove it from records to reflect new state of queue
+    if (max) {
+      delete this.urlToLink[max.url];
+    }
+
+    // Set last element to root and downheap to maintain heap property
+    if (this.links.length > 0) {
+      this.links[0] = endArray!;
+      this.downHeap(0);
+    }
+    return max;
+  }
+
+  /**
+   * Changes all priorities in queue in O(n + m*log m ) time, with n is size of urlPriorities entry
+   * and m size of the queue. Will only update priorities of urls that are actually in queue.
+   * @param urlPriorities A record with url and new priority
+   */
+  public updateAllPriority(urlPriorities: Record<string, number>): void {
+    for (const url in urlPriorities) {
+      if (this.urlToLink[url]) {
+        this.updatePriority(url, urlPriorities[url]);
+      }
+    }
+  }
+
+  /**
+   * Update priority of link in queue if it is in queue
+   * @param nodeUrl URL of link to update
+   * @param newValue new priority value
+   * @returns boolean indicating if priority was changed successfully
+   */
+  public updatePriority(nodeUrl: string, newValue: number): boolean {
+    const link = this.urlToLink[nodeUrl];
+    if (link) {
+      const change = newValue - link.metadata!.priority;
+
+      if (change === 0) {
+        return false;
+      }
+      return change > 0 ? this.increasePriority(nodeUrl, change) : this.decreasePriority(nodeUrl, -change);
+    }
+    return false;
+  }
+
+  /**
+   * Function to decrease priority of element of heap. First we check if link exists,
+   * then increase priority of the link and reheap the array.
+   */
+  public increasePriority(nodeUrl: string, increaseBy: number): boolean {
```

Comment:
`increasePriority` and `decreasePriority` look very similar.
Could we merge them into `modifyPriority` which takes a positive or negative `delta` value?

Reply:
Agreed, will change it.